core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -901,7 +907,8 @@ abstract class RDD[T: ClassTag](
    * Return an RDD that contains all matching values by applying `f`.
    */
   def collect[U: ClassTag](f: PartialFunction[T, U]): RDD[U] = withScope {
-    filter(f.isDefinedAt).map(f)
+    val cleanF = sc.clean(f)
+    filter(cleanF.isDefinedAt).map(cleanF)
```

Comment:
I believe this is correct, but I'm actually not 100% sure if it's necessary. I think it is because the `filter` closure is going to reference `f` indirectly through `f.isDefinedAt`, so it seems that we do need to clean it.

In any case, I would recommend that we keep this change since in the worst case we clone a closure twice, which is harmless (we have tests for this).